### ISSUE TITLE
False positive (?) on non-nullable property

### ISSUE THREAD

**Issue by ThomasLandauer** [phpstan/phpstan-doctrine#150]
With this being my property

```php
/**
 * @ORM\Column(type="string")
 * @Assert\NotBlank()
 */
private ?string $name = null;
```
... I'm getting this error:

> Property Foo::$name type mapping mismatch: property can contain string|null but database expects string.

So what's the correct way to set up a property that cannot be `NULL` when being saved to the database?
(It has to be nullable in PHP (i.e. `?string`) for Symfony forms to work, see https://github.com/EasyCorp/EasyAdminBundle/issues/3889#issuecomment-709924219)

**Comment by ondrejmirtes** [comment]
Mapping entities directly to forms is an antipattern, that's why you have to typehint it like this and why PHPStan doesn't like it. See: https://blog.martinhujer.cz/symfony-forms-with-request-objects/

**Comment by ThomasLandauer** [comment]
I disagree.

1. What you're essentially saying is that the entire Symfony documentation in general (and EasyAdmin as concrete implementation in particular) are doing it wrong. Even if this was true in principle (see 2), I wouldn't expect a Symfony-compatible static analyzer to throw an error on me, just for doing it the "Symfony way". In other words: I think PHPStan should concentrate on finding *errors*, not criticize the high-level architecture of the app.

2. The proposed way with custom data classes (or data transfer objects) may be a good idea in *some* situations, but certainly not in *all*. This is also acknowledged by Martin Hujer:
    > There may be exceptions. If you have some Excel-like CRUD application, where every field can be edited anytime, then it may be hard to justify using the request objects.

    If I just need a quick form for *internal use* to edit some entity, it's fine to use EasyAdmin - after all, that's what it was invented for :-)

So I think you should modify this check, or at least make it somehow user configurable.

**Comment by ondrejmirtes** [comment]
> So I think you should

This is open-source, you can send a PR ;)

But PHPStan is just telling you that your code is brittle and easily broken. If you create `new Foo()` and try to persist it + flush it to the database, you'll get the error PHPStan is telling you about.

**Comment by ThomasLandauer** [comment]
PR: Sorry, I have way too little insight :-(

`new Foo()`: What if my ultimate goal is to prevent "zombie" Foo's (i.e. with `null` as name) in the database? In this case, the non-nullable ORM mapping actually *improves* the app! Cause if somebody tries to persist an empty `Foo()`, I'd rather see the app crash than persist the zombie ;-)

Wouldn't you say this is a legitimate use case?

**Comment by ondrejmirtes** [comment]
Yes, and what PHPStan is telling you is that your app can crash during persist because of this reason.

**Comment by ThomasLandauer** [comment]
Yeah, sure. So what's your suggested solution? Just add it to `ignoreErrors`?

**Comment by ccKep** [comment]
The obvious fix would be to change your typehint from `?string` to `string`  and omit the ` = null` part.

**Comment by ondrejmirtes** [comment]
I consider this resolved.
